packages/dashboards/addon/routes/dashboards/dashboard/view.js

```diff
@@ -92,12 +92,36 @@ export default Route.extend({
      * empty filters are pruned from the request
      */
     if (cachedWidgetData && (wasEmptyFilterAdded || wasEmptyFilterRemoved)) {
-      return { dashboard, dataForWidget: cachedWidgetData };
+      return { dashboard, taskByWidget: cachedWidgetData };
     }
-    const dataForWidget = await this.get('dashboardData').fetchDataForDashboard(dashboard);
-    this.set('_widgetDataCache', dataForWidget);
 
-    return { dashboard, dataForWidget };
+    /**
+     * if we don't await here, when going back and forth between dashboards while data is still loading
+     * `deactivate` gets called after these lines, cancelling new tasks and setting the new cache to null
+     */
+    await this._cancelWidgetDataTasks();
+    const widgetsData = await this.dashboardData.fetchDataForDashboard(dashboard);
+    this.set('_widgetDataCache', widgetsData);
+
+    return { dashboard, taskByWidget: widgetsData };
+  },
+
+  /**
+   * Cancel running and enqueued widget data tasks, if any
+   *
+   * @private
+   * @method _cancelWidgetDataTasks
+   * @returns {Boolean} - true after cancelling
+   */
+  _cancelWidgetDataTasks() {
+    const { _widgetDataCache } = this;
+
+    if (!_widgetDataCache) {
+      return true;
+    }
+
+    Object.values(_widgetDataCache).forEach(widgetTask => widgetTask.cancel());
```

Comment:
Supply a message to each .cancel call saying `dashboard ${id} canceled widget task ${widgetId}`?

Reply:
done